File: skills/tokensmind-agent-network-runtime/scripts/python_runtime/http_client.py
import json
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from .errors import AgentNetworkHttpError

UNSET = object()


def _parse_response(source, status):
    if not source:
        return None
    try:
        return json.loads(source.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ValueError("Agent Network returned non-JSON HTTP %s" % status) from error


def _unwrap_response(payload, status, successful, *, headers=None):
    if successful and isinstance(payload, dict):
        if payload.get("success") is True and "data" in payload:
            return payload["data"]
        raise ValueError("Agent Network returned an invalid success response envelope")
    if successful:
        raise ValueError("Agent Network returned an invalid success response envelope")
    code = payload.get("code") if isinstance(payload, dict) else None
    message = payload.get("message") if isinstance(payload, dict) else None
    raise AgentNetworkHttpError(status, code, message, response=payload, headers=headers)


class HttpClient:
    def __init__(self, opener=urlopen):
        self._opener = opener

    def request(self, url, method="GET", headers=None, *, body=UNSET):
        data = None
        if body is not UNSET:
            data = json.dumps(body, separators=(",", ":")).encode("utf-8")
        request = Request(url, data=data, headers=headers or {}, method=method)
        try:
            response = self._opener(request)
            with response:
                status = response.getcode()
                headers = {"retryAfter": response.headers.get("Retry-After")}
                payload = _parse_response(response.read(), status)
            return _unwrap_response(payload, status, 200 <= status < 300, headers=headers)
        except HTTPError as error:
            headers = {"retryAfter": error.headers.get("Retry-After")}
            payload = _parse_response(error.read(), error.code)
            return _unwrap_response(payload, error.code, False, headers=headers)

File: skills/tokensmind-agent-network-runtime/scripts/lib/runtime/httpClient.js
export class AgentNetworkHttpError extends Error {
  constructor({ status, code, message, response, retryAfter, correction }) {
    super(message || `Agent Network request failed with HTTP ${status}`);
    this.name = 'AgentNetworkHttpError';
    this.status = status;
    this.code = code;
    this.response = response;
    this.retryAfter = retryAfter;
    this.correction = correction;
  }
}

async function parseResponse(response) {
  const text = await response.text();
  if (!text) return null;
  try {
    return JSON.parse(text);
  } catch {
    throw new Error(`Agent Network returned non-JSON HTTP ${response.status}`);
  }
}

function unwrapSuccess(response, payload) {
  if (!response.ok) return undefined;
  if (payload?.success === true && Object.hasOwn(payload, 'data')) return payload.data;
  throw new Error('Agent Network returned an invalid success response envelope');
}

function throwHttpError(response, payload) {
  throw new AgentNetworkHttpError({
    status: response.status,
    code: payload?.code,
    message: payload?.message,
    response: payload,
    retryAfter: response.headers.get('Retry-After'),
    correction: payload?.correction,
  });
}

export function createHttpClient({ fetchImpl = fetch } = {}) {
  return {
    async request({ url, method = 'GET', headers, body, signal }) {
      const response = await fetchImpl(url, {
        method,
        headers,
        body: body === undefined ? undefined : JSON.stringify(body),
        signal,
      });
      const payload = await parseResponse(response);
      if (response.ok) return unwrapSuccess(response, payload);
      return throwHttpError(response, payload);
    },
  };
}
